File: docs/JAMIA/comparisons/RPython/TM/22557322_savetxt_fmt_drops_zeros.ts
/**
 * ID: SO#22557322
 * Language: Python
 * Bug class: Value type
 * Runtime consequence: IF
 * In study: Yes
 * Inclusion rationale: numpy savetxt fmt='%i' on float array silently drops zeroes. Format/type mismatch produces wrong output.
 */
import { createDataFrame, stats as s } from "@tidy-ts/dataframe";
import { printForeignResult, runForeign } from "../run-foreign.ts";

// Foreign reproduction (pandas) ──────────────────────────────────────────────

const foreignScript = `
import numpy as np
import io

result = np.array([
    [1.0, 2.0],
    [2.0, 0.0],
    [3.0, 9.0],
    [4.0, 0.0],
    [5.0, 3.0],
])

buf = io.BytesIO()
np.savetxt(buf, result, fmt='%i', delimiter=',')
print(buf.getvalue().decode())
`;

printForeignResult("python", runForeign("python", foreignScript));

// Tidy-TS equivalent ─────────────────────────────────────────────────────────

const df = createDataFrame([
  { col1: 1.0, col2: 2.0 },
  { col1: 2.0, col2: 0.0 },
  { col1: 3.0, col2: 9.0 },
  { col1: 4.0, col2: 0.0 },
  { col1: 5.0, col2: 3.0 },
]);

const formatted = df.mutate({ col2_str: (r) => r.col2.toFixed(0) });

// @ts-expect-error — Argument of type 'string[]' is not assignable to parameter of type 'NumbersWithNullable'.
s.sum(formatted.extract("col2_str"));

File: docs/JAMIA/comparisons/RPython/TM/22557322_savetxt_fmt_drops_zeros.py
# ID: SO#22557322
# Language: Python
# Bug class: Value type
# Runtime consequence: IF
# In study: Yes
# Inclusion rationale: numpy savetxt fmt='%i' on float array silently drops zeroes. Format/type mismatch produces wrong output.
# Reproduction status: Fixed (modern numpy)
# Type system catch: Format specifier `%i` rejects `float` array at type level

import numpy as np
import io

result = np.array([
    [1.0, 2.0],
    [2.0, 0.0],
    [3.0, 9.0],
    [4.0, 0.0],
    [5.0, 3.0],
])

buf = io.BytesIO()
np.savetxt(buf, result, fmt='%i', delimiter=',')
output = buf.getvalue().decode()
print("Output with fmt='%i':")
print(output)
# Note: zero values appear as "0" which is correct on modern numpy,
# but the original bug had them disappear entirely
